# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1765910772 -3600
# Parent  34830809897efe63624fb6c0fe8d11e561b61ba1
Feature: Implement dark reader switch for the compose window.

diff --git a/mail/app/profile/all-thunderbird.js b/mail/app/profile/all-thunderbird.js
--- a/mail/app/profile/all-thunderbird.js
+++ b/mail/app/profile/all-thunderbird.js
@@ -1555,8 +1555,9 @@ pref("mail.thread.conversation.enabled",
 #endif
 
 // Enable the conversion to dark mode for all messages when using a dark theme.
 pref("mail.dark-reader.enabled", true);
+pref("mail.dark-reader-compose.enabled", true);
 pref("mail.dark-reader.show-toggle", true);
 
 // Enable the new account setup (starting from the second account)
 #ifdef NIGHTLY_BUILD
diff --git a/mail/components/compose/content/MsgComposeCommands.js b/mail/components/compose/content/MsgComposeCommands.js
--- a/mail/components/compose/content/MsgComposeCommands.js
+++ b/mail/components/compose/content/MsgComposeCommands.js
@@ -5970,8 +5970,31 @@ async function ComposeLoad() {
   ];
 
   UIDensity.registerWindow(window);
   UIFontSize.registerWindow(window);
+
+  // Wire up the dark reader toggle.
+  const disableDarkReaderToggle = document.getElementById("disableDarkReaderCompose");
+  disableDarkReaderToggle.checked = !Services.prefs.getBoolPref(
+    "mail.dark-reader-compose.enabled",
+    true
+  );
+  document.l10n.setAttributes(
+    disableDarkReaderToggle,
+    disableDarkReaderToggle.checked
+      ? "dark-message-mode-toggle-disabled"
+      : "dark-message-mode-toggle-enabled"
+  );
+  disableDarkReaderToggle.addEventListener("click", e => {
+    Services.prefs.setBoolPref("mail.dark-reader-compose.enabled", !e.target.checked);
+    document.l10n.setAttributes(
+      disableDarkReaderToggle,
+      disableDarkReaderToggle.checked
+        ? "dark-message-mode-toggle-disabled"
+        : "dark-message-mode-toggle-enabled"
+    );
+    focusMsgBody();
+  });
 }
 
 /**
  * Add fluent strings to labels and menu items requiring a shortcut key.
diff --git a/mail/components/compose/content/messengercompose.xhtml b/mail/components/compose/content/messengercompose.xhtml
--- a/mail/components/compose/content/messengercompose.xhtml
+++ b/mail/components/compose/content/messengercompose.xhtml
@@ -58,8 +58,9 @@
   <link rel="stylesheet" href="chrome://messenger/skin/inContentDialog.css" />
 
   <link rel="localization" href="branding/brand.ftl"/>
   <link rel="localization" href="messenger/messenger.ftl" />
+  <link rel="localization" href="messenger/messageheader/headerFields.ftl"/>
   <link rel="localization" href="messenger/messengercompose/messengercompose.ftl" />
   <link rel="localization" href="messenger/menubar.ftl" />
   <link rel="localization" href="messenger/appmenu.ftl" />
   <link rel="localization" href="messenger/openpgp/openpgp.ftl" />
@@ -2445,8 +2446,15 @@
                persist="collapsed"
                nowindowdrag="true">
 #include editFormatButtons.inc.xhtml
         <spacer flex="1"/>
+        <html:div id="darkReaderToggleCompose" xmlns="http://www.w3.org/1999/xhtml">
+          <label for="disableDarkReaderCompose" class="toggle-label">
+            <input id="disableDarkReaderCompose"
+                   type="checkbox"
+                   class="toggle-checkbox toggle-checkbox-sm dark-reader-toggle" />
+          </label>
+        </html:div>
       </toolbar>
     </toolbox>
 
     <html:hr is="pane-splitter" id="headersSplitter"
diff --git a/mail/themes/shared/mail/editorContent.css b/mail/themes/shared/mail/editorContent.css
--- a/mail/themes/shared/mail/editorContent.css
+++ b/mail/themes/shared/mail/editorContent.css
@@ -97,9 +97,9 @@ img {
   }
 }
 
 /* Can be removed when it is in messageQuotes.css enabled again */
-@media (prefers-color-scheme: dark) {
+@media (prefers-color-scheme: dark) and -moz-pref("mail.dark-reader-compose.enabled") {
   body {
     color: #f9f9fa;
     background-color: #2a2a2e;
   }
diff --git a/mail/themes/shared/mail/messengercompose.css b/mail/themes/shared/mail/messengercompose.css
--- a/mail/themes/shared/mail/messengercompose.css
+++ b/mail/themes/shared/mail/messengercompose.css
@@ -1662,4 +1662,117 @@ dialog .key-info-block {
 
 .comma-separated a:not(:last-child):after {
   content: ", ";
 }
+
+#darkReaderToggleCompose {
+  display: inline-block;
+  @media not (prefers-color-scheme: dark) {
+    display: none;
+  }
+}
+
+/* copied from widgets.css */
+.toggle-checkbox {
+  appearance: none;
+  box-sizing: border-box;
+  padding: 0;
+  background-color: var(--color-surface-lower);
+  /* BB mod: border: 1px solid var(--color-surface-border); */
+  block-size: 22px;
+  inline-size: 44px;
+  border-radius: 14px;
+  cursor: pointer;
+
+  &::before {
+    display: block;
+    content: "";
+    box-sizing: border-box;
+    /* BB mod: background-color: var(--color-surface-base); */
+    background-color: color-mix(in srgb, var(--selected-item-text-color) 40%, transparent);
+    border: 1px solid var(--color-surface-border);
+    block-size: 18px;
+    inline-size: 18px;
+    margin-block-start: 1px;
+    margin-inline-start: 1px;
+    border-radius: 10px;
+    box-shadow: 0 1px 2px rgba(0, 0, 0, 0.1);
+    translate: 0;
+  }
+
+  &:checked {
+    background-color: var(--selected-item-color);
+    border-color: color-mix(in srgb, var(--selected-item-color) 80%, black);
+
+    &::before {
+      translate: 23px;
+      background-color: var(--selected-item-text-color);
+      border-color: color-mix(in srgb, var(--selected-item-color) 80%, black);
+    }
+
+    &:is(:hover, :active) {
+      background-color: color-mix(in srgb, var(--selected-item-color) 90%, black);
+      border-color: color-mix(in srgb, var(--selected-item-color) 80%, black);
+    }
+  }
+
+  /* BB mod:
+  &:is(:hover, :active) {
+    background-color: var(--color-surface-border);
+    border-color: var(--color-neutral-200);
+  }
+  */
+
+  &:focus-visible {
+    outline: var(--focus-outline);
+    outline-offset: var(--focus-outline-offset);
+  }
+
+  &.toggle-checkbox-sm {
+    block-size: 18px;
+    inline-size: 38px;
+
+    &::before {
+      block-size: 14px;
+      inline-size: 14px;
+    }
+
+    &:checked::before {
+      translate: 20px;
+    }
+  }
+}
+
+/* copied from messageHeader.css */
+.dark-reader-toggle {
+  position: relative;
+
+  &::after {
+    position: absolute;
+    display: flex;
+    align-items: center;
+    content: var(--icon-moon-xs);
+    -moz-context-properties: fill, stroke;
+    fill: color-mix(in srgb, currentColor 10%, transparent);
+    stroke: color-mix(in srgb, currentColor 55%, transparent);
+    inset-block: 0;
+    inset-inline-end: 3px;
+  }
+
+  &:checked::after {
+    content: var(--icon-sun-xs);
+    fill: color-mix(in srgb, var(--selected-item-text-color) 40%, transparent);
+    stroke: color-mix(in srgb, var(--selected-item-text-color) 90%, transparent);
+    translate: -23px;
+  }
+
+  &.toggle-checkbox-sm {
+    &::after {
+      inset-inline-end: 2px;
+    }
+
+    &:checked::after {
+      translate: -20px;
+    }
+  }
+}
+
